File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findNumberOfLIS = function(nums) {

    n = nums.length;
    const dpLis  = Array(n).fill(1);
    const dpLisCount = Array(n).fill(1);
    
    for(let i=0 ;i<n;i++){

        let j= i-1; 
        while(j>=0){
            if(nums[j]< nums[i]){
                if(dpLis[i]<1+dpLis[j]){

                    dpLis[i] = 1+dpLis[j]
                    dpLisCount[i]=0
                    dpLisCount[i] += dpLisCount[j]


                }
                else if(dpLis[i]===1+dpLis[j]){
                    dpLisCount[i] +=dpLisCount[j]

                }
           
            }
            
            j-=1
        }
       

    }
    // console.log(dpLis,dpLisCount)
    return dpLisCount.reduce((acc, val,ind)=>{

                if (dpLis[ind] === Math.max(...dpLis)){

                    return acc+ val;
                }
                else{
                    return acc;
                }


    },0 );


    
    
};

File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:

        n= len(nums)
        dp =[1 for i in range(n)]
        cnt = [1 for i in range(n)]
  
        ind=0

        #FIND lenght of LIS
        max_ =1 
        while(ind<n):
            prevInd=ind-1
            while(prevInd>-1 ):
                if nums[prevInd]<nums[ind]:
                    if dp[prevInd]+1> dp[ind]:
                        dp[ind] =dp[prevInd]+1
                        cnt[ind] = cnt[prevInd]
                    elif dp[prevInd]+1== dp[ind]:
                        cnt[ind]+= cnt[prevInd]



                prevInd-=1
            max_ = max(max_, dp[ind])
            ind+=1
        ans  =0 
        print(dp, cnt)
        for i in range(n):
            if dp[i] == max_:
                ans+= cnt[i]
        return ans



